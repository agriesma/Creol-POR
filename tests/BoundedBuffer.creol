interface BufferA
begin with Producer
  op append(in d: Data)
end

interface BufferR
begin with Consumer
  op remove(out d: Data)
end

interface Buffer inherits BufferA, BufferB
begin 
end

class BoundedBuffer implements Buffer
begin 
  var buffer: List := empty
  var max: Int := 10
  var n: Int := 0

  with Producer
    op append(in d:Data) == await n < max; buffer := plus(buffer, d);
  with Consumer
    op remove(out d:Data) ==
      await n > 0; d := first(buffer); buffer := rest(buffer);
end


interface Consumer
begin
end

interface Producer
begin 
end

class Prod(b: BufferA)
  implements Producer
begin
  op run == !this.loop(0);
  op loop(in i: Int) == b.append(i;); !this.loop(i+1);
end


class Cons(b:BufferR) implements Consumer
begin 
  op run == !this.loop();
  op loop == var y:Data; b.remove(;y); /* print(y;); */ !this.loop();
end

class Starter
begin 
  op run == var b: Buffer; var p: Producer; var c: Consumer; 
            b := new BoundedBuffer(); 
            p := new Producer(b);
            c := new Consumer(b);
end
