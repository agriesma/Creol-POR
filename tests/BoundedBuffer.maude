load interpreter
mod PROGRAM is
pr INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< 'BoundedBuffer : Cl | Inh: noInh, Par: noAid, Att: 'buffer |-> null , 'max |-> null , 'n |-> null, Mtds: 
  < 'append : Mtdname | Param: 'd, Latt: 'd |-> null, Code: skip ; skip ; ( await ( 'less[[ 'n # 'max ]] ) ) ; 'buffer ::= ( 'plus[[ 'buffer # 'd ]] ) ; return ( emp ) > *
  < 'remove : Mtdname | Param: noAid, Latt: 'd |-> null, Code: skip ; skip ; ( await ( 'less[[ int(0) # 'n ]] ) ) ; 'd ::= ( 'first[[ 'buffer ]] ) ; 'buffer ::= ( 'rest[[ 'buffer ]] ) ; return ( 'd ) >, Ocnt: 0 >

< 'Prod : Cl | Inh: noInh, Par: 'b, Att: 'b |-> null, Mtds: 
  < 'run : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; skip ; 'this ! 'this . 'loop ( int(0) ) ; return ( emp ) > *
  < 'loop : Mtdname | Param: 'i, Latt: 'i |-> null, Code: skip ; skip ; 'b . 'append( 'i : noAid ) ;  ! 'this . 'loop ( ( 'plus[[ 'i # int(1) ]] ) ) ; return ( emp ) >, Ocnt: 0 >

< 'Cons : Cl | Inh: noInh, Par: 'b, Att: 'b |-> null, Mtds: 
  < 'run : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; skip ;  ! 'this . 'loop ( emp ) ; return ( emp ) > *
  < 'loop : Mtdname | Param: noAid, Latt: 'y |-> null, Code: skip ; skip ; 'b . 'remove( emp : 'y ) ;  ! 'this . 'loop ( emp ) ; return ( emp ) >, Ocnt: 0 >

< 'Starter : Cl | Inh: noInh, Par: noAid, Att: noSubst, Mtds: 
  < 'run : Mtdname | Param: noAid, Latt: 'b |-> null , 'p |-> null , 'c |-> null, Code: skip ; skip ; 'b ::= new 'BoundedBuffer( emp ) ; 'p ::= new 'Producer( 'b ) ; 'c ::= new 'Consumer( 'b ) ; return ( emp ) >, Ocnt: 0 >

.
endm
