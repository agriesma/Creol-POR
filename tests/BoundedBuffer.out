==========================================
reduce in PROGRAM : init .
result Configuration: 
< "BoundedBuffer" : Cl | Inh: noInh, Par: noAid, Att: "buffer" |-> null, "max"
    |-> null, "n" |-> null, Mtds:
  < "append" : Mtdname | Param: "d", Latt: "d" |-> null, Code: await "<"("n" #
    "max") ; "buffer" ::= "plus"("buffer" # "d") ; return(emp) > *
  < "remove" : Mtdname | Param: noAid, Latt: "d" |-> null, Code: await ">"("n"
    # int(0)) ; "d" ::= "first"("buffer") ; "buffer" ::= "rest"("buffer") ;
    return("d") >,
  Ocnt: 0 >

< "Cons" : Cl | Inh: noInh, Par: "b", Att: "b" |-> null, Mtds:
  < "loop" : Mtdname | Param: noAid, Latt: "y" |-> null, Code: "sync.4" ! "b" .
    "remove"(emp) ; "sync.4" ?("y") ; "anon.5" ! "this" . "loop"(emp) ; return(
    emp) > *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: "anon.3" ! "this" .
    "loop"(emp) ; return(emp) >,
  Ocnt: 0 >

< "Prod" : Cl | Inh: noInh, Par: "b", Att: "b" |-> null, Mtds:
  < "loop" : Mtdname | Param: "i", Latt: "i" |-> null, Code: "sync.1" ! "b" .
    "append"("i") ; "sync.1" ?(noAid) ; "anon.2" ! "this" . "loop"("+"("i" #
    int(1))) ; return(emp) > *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: "anon.0" ! "this" .
    "loop"(int(0)) ; return(emp) >,
  Ocnt: 0 >

< "Starter" : Cl | Inh: noInh, Par: noAid, Att: noSubst, Mtds:
  < "run" : Mtdname | Param: noAid, Latt: "b" |-> null, "c" |-> null, "p" |->
    null, Code: "b" ::= new "BoundedBuffer"(emp) ; "p" ::= new "Producer"("b")
    ; "c" ::= new "Consumer"("b") ; return(emp) >,
  Ocnt: 0 >
Bye.
